File: web/src/types/theme.ts
/**
 * Theme Types
 *
 * TypeScript interfaces for theme analysis entities.
 * These types mirror the backend Python models in src/models/theme.py
 *
 * Theme analysis identifies patterns and topics across multiple content items,
 * tracking how themes evolve over time using the knowledge graph.
 *
 * @see Backend model: src/models/theme.py
 */

/**
 * Theme categories for classification
 *
 * Categories help organize themes into meaningful groups
 * for filtering and navigation.
 */
export type ThemeCategory =
  | "ml_ai" // Machine learning and AI topics
  | "devops_infra" // DevOps, infrastructure, cloud
  | "data_engineering" // Data pipelines, warehousing
  | "business_strategy" // Business and strategy topics
  | "tools_products" // New tools, product launches
  | "research_academia" // Research papers, academic work
  | "security" // Security, privacy, compliance
  | "other" // Uncategorized

/**
 * Theme trend classification
 *
 * Indicates how the theme is evolving over time
 * based on historical mentions in the knowledge graph.
 */
export type ThemeTrend =
  | "emerging" // New topic, first mentions recently
  | "growing" // Increasing frequency of mentions
  | "established" // Consistent, stable presence
  | "declining" // Decreasing frequency
  | "one_off" // Single mention, not recurring

/**
 * Entity from the knowledge graph
 *
 * Represents a node in the Neo4j knowledge graph,
 * extracted and maintained by Graphiti.
 */
export interface Entity {
  /** Unique entity identifier */
  id: string
  /** Entity name/label */
  name: string
  /** Entity type (e.g., "Technology", "Company", "Person") */
  type: string
  /** Additional properties/attributes */
  properties: Record<string, unknown>
}

/**
 * Relationship between entities
 *
 * Represents an edge in the knowledge graph
 * connecting two entities.
 */
export interface Relationship {
  /** Unique relationship identifier */
  id: string
  /** Source entity ID */
  sourceId: string
  /** Target entity ID */
  targetId: string
  /** Relationship type (e.g., "USES", "COMPETES_WITH") */
  type: string
  /** Confidence score (0-1) */
  confidence: number
  /** Additional properties */
  properties: Record<string, unknown>
}

/**
 * Individual theme within an analysis
 *
 * Represents a single identified theme with
 * its classification and supporting data.
 */
export interface Theme {
  /** Theme name/title */
  name: string

  /** Category classification */
  category: ThemeCategory

  /** Trend based on historical data */
  trend: ThemeTrend

  /** Confidence score (0-1) of theme identification */
  confidence: number

  /** Number of mentions across analyzed content items */
  mentions: number

  /** IDs of content items that mention this theme */
  sourceContentIds: string[]

  /** Brief summary of the theme */
  summary: string

  /** Related entities from knowledge graph */
  relatedEntities: Entity[]

  /** Key quotes or excerpts about this theme */
  keyExcerpts: string[]

  /**
   * Historical context from knowledge graph
   * Previous mentions and evolution of this theme
   */
  historicalContext?: {
    /** When theme was first seen */
    firstMention: string
    /** Total historical mentions */
    totalMentions: number
    /** Narrative of how theme has evolved */
    evolutionNarrative: string
  }
}

/**
 * Theme Analysis entity
 *
 * Results of analyzing themes across a set of content items
 * for a specific time period.
 */
export interface ThemeAnalysis {
  /** Unique identifier (UUID) */
  id: string

  /** Date the analysis was performed */
  analysisDate: string // ISO 8601

  /** Start of the analyzed period */
  startDate: string // ISO 8601

  /** End of the analyzed period */
  endDate: string // ISO 8601

  /** Number of content items analyzed */
  contentCount: number

  /** IDs of analyzed content items */
  contentIds: string[]

  /** Identified themes */
  themes: Theme[]

  /** Total number of themes identified */
  totalThemes: number

  /** Count of emerging themes */
  emergingThemesCount: number

  /** Name of the top/most prominent theme */
  topTheme: string

  /** Agent framework used */
  agentFramework: string

  /** Model used for analysis */
  modelUsed: string

  /** Model version */
  modelVersion: string | null

  /** Processing time in seconds */
  processingTimeSeconds: number

  /** Tokens used */
  tokenUsage: number
}

/**
 * Graph data structure for visualization
 *
 * Format suitable for react-force-graph and similar libraries.
 */
export interface GraphData {
  /** Nodes in the graph */
  nodes: GraphNode[]
  /** Links/edges in the graph */
  links: GraphLink[]
}

/**
 * Node in the graph visualization
 */
export interface GraphNode {
  /** Unique node ID */
  id: string
  /** Display label */
  label: string
  /** Node type for coloring/grouping */
  type: "theme" | "entity" | "content"
  /** Optional size weight */
  weight?: number
  /** Category for themes */
  category?: ThemeCategory
  /** Trend for themes */
  trend?: ThemeTrend
  /** Additional display properties */
  properties?: Record<string, unknown>
}

/**
 * Link/edge in the graph visualization
 */
export interface GraphLink {
  /** Source node ID */
  source: string
  /** Target node ID */
  target: string
  /** Relationship type label */
  type: string
  /** Link strength/weight */
  weight?: number
}

/**
 * Request to trigger theme analysis
 */
export interface AnalyzeThemesRequest {
  /** Start date for analysis period */
  startDate: string
  /** End date for analysis period */
  endDate: string
  /** Specific content IDs (if empty, uses date range) */
  contentIds?: string[]
  /** Include historical context from knowledge graph */
  includeHistoricalContext?: boolean
}

/**
 * Filters for theme analysis list
 */
export interface ThemeAnalysisFilters {
  /** Filter by date range start */
  startDate?: string
  /** Filter by date range end */
  endDate?: string
  /** Minimum number of themes */
  minThemes?: number
  /** Pagination limit */
  limit?: number
  /** Pagination offset */
  offset?: number
}

/**
 * Theme data from backend ThemeData model
 */
export interface ThemeData {
  name: string
  description: string
  category: ThemeCategory
  mention_count: number
  content_ids: number[]
  first_seen: string
  last_seen: string
  trend: ThemeTrend
  relevance_score: number
  strategic_relevance: number
  tactical_relevance: number
  novelty_score: number
  cross_functional_impact: number
  related_themes: string[]
  key_points: string[]
  historical_context?: {
    theme_name: string
    first_mention: string
    total_mentions: number
    mention_frequency: string
    evolution_summary: string
    previous_discussions: string[]
    stance_change?: string
    recent_mentions: Array<{
      date: string
      content_id: number
      content_title: string
      publication: string
      context: string
      sentiment?: string
    }>
  }
  continuity_text?: string
}

/**
 * Result from theme analysis (matches backend ThemeAnalysisResult)
 */
export interface ThemeAnalysisResult {
  id: number
  analysis_date: string
  start_date: string
  end_date: string
  content_count: number
  content_ids: number[]
  themes: ThemeData[]
  total_themes: number
  emerging_themes_count: number
  top_theme: string | null
  processing_time_seconds: number
  token_usage: number | null
  model_used: string
  model_version: string | null
  agent_framework: string
  cross_theme_insights: string[]
  error_message?: string
  created_at?: string
}

File: src/models/theme.py
"""Theme analysis data models."""

from datetime import UTC, datetime
from enum import StrEnum

from pydantic import BaseModel, Field
from sqlalchemy import JSON, Column, DateTime, Enum, Float, Index, Integer, String, Text

from src.models.base import Base


class ThemeCategory(StrEnum):
    """Theme categorization."""

    ML_AI = "ml_ai"  # LLMs, agents, RAG, fine-tuning
    DEVOPS_INFRA = "devops_infra"  # Cloud, containers, orchestration
    DATA_ENGINEERING = "data_engineering"  # Pipelines, lakes, processing
    BUSINESS_STRATEGY = "business_strategy"  # Adoption, ROI, governance
    TOOLS_PRODUCTS = "tools_products"  # New releases, updates
    RESEARCH_ACADEMIA = "research_academia"  # Papers, breakthroughs
    SECURITY = "security"  # Security, privacy, compliance
    OTHER = "other"  # Miscellaneous


class ThemeTrend(StrEnum):
    """Theme trend classification."""

    EMERGING = "emerging"  # New topic, recent mentions
    GROWING = "growing"  # Increasing frequency
    ESTABLISHED = "established"  # Consistent mentions over time
    DECLINING = "declining"  # Decreasing mentions
    ONE_OFF = "one_off"  # Single mention


class AnalysisStatus(StrEnum):
    """Theme analysis lifecycle status."""

    QUEUED = "queued"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class ThemeAnalysis(Base):
    """Theme analysis results database model."""

    __tablename__ = "theme_analyses"

    id = Column(Integer, primary_key=True, autoincrement=True)

    # Lifecycle status
    status = Column(
        Enum(
            AnalysisStatus,
            name="analysisstatus",
            create_type=False,
            create_constraint=True,
            values_callable=lambda x: [e.value for e in x],
        ),
        nullable=False,
        default=AnalysisStatus.QUEUED,
        index=True,
    )

    # Time range for analysis
    analysis_date = Column(DateTime, nullable=False, default=lambda: datetime.now(UTC), index=True)
    start_date = Column(DateTime, nullable=False)
    end_date = Column(DateTime, nullable=False)

    # Analysis scope (renamed from newsletter_*)
    content_count = Column(Integer, nullable=False, default=0)
    content_ids = Column(JSON, nullable=False, default=list)  # List[int]

    # Detected themes
    themes = Column(JSON, nullable=False, default=list)  # List[ThemeData dict]

    # Summary statistics
    total_themes = Column(Integer, nullable=False, default=0)
    emerging_themes_count = Column(Integer, nullable=False, default=0)
    top_theme = Column(String(500), nullable=True)

    # Metadata
    agent_framework = Column(String(100), nullable=False, default="")
    model_used = Column(String(100), nullable=False, default="")
    model_version = Column(String(20), nullable=True)
    processing_time_seconds = Column(Float, nullable=True)
    token_usage = Column(Integer, nullable=True)

    # New persistence fields
    cross_theme_insights = Column(JSON, nullable=True)  # List[str]
    error_message = Column(Text, nullable=True)
    created_at = Column(DateTime, nullable=False, default=lambda: datetime.now(UTC), index=True)

    __table_args__ = (Index("ix_theme_analyses_status_created", "status", "created_at"),)


class HistoricalMention(BaseModel):
    """Historical mention of a theme."""

    date: datetime = Field(..., description="Date of mention")
    newsletter_id: int = Field(..., description="Content ID (legacy field name)")
    newsletter_title: str = Field(..., description="Content title (legacy field name)")
    publication: str = Field(..., description="Publication name")
    context: str = Field(..., description="Context snippet about the theme")
    sentiment: str | None = Field(
        None, description="Sentiment/stance (positive, neutral, negative)"
    )


class ThemeEvolution(BaseModel):
    """How a theme has evolved over time."""

    theme_name: str = Field(..., description="Theme name")
    first_mention: datetime = Field(..., description="When first discussed")
    total_mentions: int = Field(..., description="Total historical mentions")
    mention_frequency: str = Field(
        ..., description="Frequency (rare, occasional, frequent, constant)"
    )

    # Evolution narrative
    evolution_summary: str = Field(..., description="How the theme has evolved (1-2 sentences)")
    previous_discussions: list[str] = Field(
        default_factory=list, description="Key points from previous discussions"
    )

    # Change tracking
    stance_change: str | None = Field(
        None, description="How stance/sentiment has changed over time"
    )

    # Historical mentions
    recent_mentions: list[HistoricalMention] = Field(
        default_factory=list, description="Recent mentions (last 3-5)"
    )


class ThemeData(BaseModel):
    """Individual theme data."""

    name: str = Field(..., description="Theme name (e.g., 'RAG Architecture')")
    description: str = Field(..., description="Brief description of the theme")
    category: ThemeCategory = Field(..., description="Theme category")

    # Frequency and recency
    mention_count: int = Field(..., description="Number of content items mentioning this theme")
    content_ids: list[int] = Field(
        default_factory=list,
        description="IDs of content items mentioning this theme",
    )
    first_seen: datetime = Field(..., description="First mention date")
    last_seen: datetime = Field(..., description="Most recent mention date")

    # Trend classification
    trend: ThemeTrend = Field(..., description="Theme trend (emerging, growing, etc.)")

    # Relevance scoring
    relevance_score: float = Field(..., description="Overall relevance (0-1)")
    strategic_relevance: float = Field(..., description="Strategic/leadership relevance (0-1)")
    tactical_relevance: float = Field(..., description="Tactical/developer relevance (0-1)")
    novelty_score: float = Field(..., description="How novel vs. established (0-1)")
    cross_functional_impact: float = Field(..., description="Cross-team impact (0-1)")

    # Related themes
    related_themes: list[str] = Field(default_factory=list, description="Names of related themes")

    # Key insights
    key_points: list[str] = Field(
        default_factory=list, description="Key points about this theme from content items"
    )

    # Historical context
    historical_context: ThemeEvolution | None = Field(
        None, description="Historical context and evolution of this theme"
    )

    # Continuity text
    continuity_text: str | None = Field(
        None, description="Human-readable continuity statement (e.g., 'Previously discussed in...')"
    )


class ThemeAnalysisRequest(BaseModel):
    """Request parameters for theme analysis."""

    start_date: datetime
    end_date: datetime
    min_newsletters: int = Field(default=1, description="Minimum content items to analyze")
    max_themes: int = Field(default=20, description="Maximum themes to return")
    relevance_threshold: float = Field(default=0.3, description="Minimum relevance score (0-1)")
    use_large_context_model: bool = Field(
        default=False, description="Use large context model (Gemini Flash) for analysis"
    )


class ThemeAnalysisResult(BaseModel):
    """Complete theme analysis results."""

    id: int | None = None
    analysis_date: datetime = Field(default_factory=lambda: datetime.now(UTC))
    start_date: datetime
    end_date: datetime

    # Content items analyzed (renamed from newsletter_*)
    content_count: int = 0
    content_ids: list[int] = Field(default_factory=list)

    # Themes
    themes: list[ThemeData] = Field(default_factory=list)
    total_themes: int = 0
    emerging_themes_count: int = 0

    # Top theme
    top_theme: str | None = None

    # Performance metrics
    processing_time_seconds: float = 0.0
    token_usage: int | None = None
    model_used: str = ""  # General model ID
    model_version: str | None = None  # Version
    agent_framework: str = ""

    # Insights
    cross_theme_insights: list[str] = Field(
        default_factory=list, description="Insights about connections between themes"
    )
